Objective-C/CBLArray.m

```diff
@@ -0,0 +1,342 @@
+//
+//  CBLArraym
+//  CouchbaseLite
+//
+//  Created by Pasin Suriyentrakorn on 4/12/17.
+//  Copyright © 2017 Couchbase. All rights reserved.
+//
+
+#import "CBLArray.h"
+#import "CBLBlob.h"
+#import "CBLData.h"
+#import "CBLDocument+Internal.h"
+#import "CBLJSON.h"
+#import "CBLSubdocument.h"
+
+@implementation CBLArray {
+    NSMutableArray* _array;
+    NSMapTable* _changeListeners;
+    BOOL _changed;
+}
+
+
++ (instancetype) array {
+    return [[self alloc] init];
+}
+
+
+- (instancetype) init {
+    return [self initWithFleeceData: nil];
+}
+
+
+- (instancetype) initWithArray: (NSArray*)array {
+    self = [self init];
+    if (self) {
+        [self setArray: array];
+    }
+    return self;
+}
+
+
+- /*internal */ (instancetype) initWithFleeceData: (CBLFLArray*)data {
+    self = [super initWithFleeceData: data];
+    if (self ) {
+        _array = [NSMutableArray array];
+        [self loadBackingData];
+    }
+    return self;
+}
+
+
+#pragma mark - GETTER
+
+
+- (nullable id) objectAtIndex: (NSUInteger)index {
+    return _array[index];
+}
+
+
+- (BOOL) booleanAtIndex: (NSUInteger)index {
+    id value = _array[index];
+    return [CBLData booleanValueForObject: value];
```

Comment:
Current boolean has a special logic to convert the value. If the value is not NSNumber and NSNull, it will return true. I this the same LiteCore boolean conversion behavior. I will check with Jens the reason behind this logic. I could guess that this is inline with if (obj) in iOS.